iot_packet_verifier/src/daemon.rs

```diff
@@ -0,0 +1,215 @@
+use crate::{balances::Balances, burner::Burner, ingest, settings::Settings};
+use anyhow::{Error, Result};
+use chrono::{DateTime, NaiveDateTime, Utc};
+use db_store::meta;
+use file_store::{file_sink, file_sink_write, file_upload, FileStore, FileType};
+use futures::StreamExt;
+use futures_util::TryFutureExt;
+use helium_crypto::{PublicKey, PublicKeyBinary};
+use helium_proto::services::{
+    iot_config::{org_client::OrgClient, OrgGetReqV1},
+    packet_verifier::ValidPacket,
+    Channel,
+};
+use solana_client::nonblocking::rpc_client::RpcClient;
+use solana_sdk::pubkey::Pubkey;
+use sqlx::{Pool, Postgres};
+use std::{collections::HashMap, sync::Arc};
+use tokio::time;
+
+struct Daemon {
+    pool: Pool<Postgres>,
+    file_store: FileStore,
+    balances: Balances,
+    org_client: OrgClient<Channel>,
+    sub_dao: Pubkey,
+    valid_packets_tx: file_sink::MessageSender,
+    invalid_packets_tx: file_sink::MessageSender,
+}
+
+const POLL_TIME: time::Duration = time::Duration::from_secs(10);
+
+impl Daemon {
+    pub async fn run(mut self, shutdown: &triggered::Listener) -> Result<()> {
+        let mut timer = time::interval(POLL_TIME);
+        timer.set_missed_tick_behavior(time::MissedTickBehavior::Delay);
+
+        let mut last_verified: NaiveDateTime =
+            meta::fetch(&self.pool, "last_verified_packet").await?;
+
+        loop {
+            tokio::select! {
+                _ = shutdown.clone() => break,
+                _ = timer.tick() => {
+                    last_verified = self.handle_tick(last_verified.clone()).await?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn handle_tick(&mut self, mut last_verified: NaiveDateTime) -> Result<NaiveDateTime> {
+        let reports =
+            ingest::ingest_reports(&self.file_store, DateTime::from_utc(last_verified, Utc));
+
+        tokio::pin!(reports);
+
+        let mut org_cache = HashMap::<u64, PublicKey>::new();
+
+        while let Some(report) = reports.next().await {
+            let report_timestamp =
+                NaiveDateTime::from_timestamp_millis(report.gateway_timestamp_ms as i64).unwrap();
+
+            // Since the report timestamp will always be behind the timestamp for the while,
+            // the first few reports we see upon restarting will have already been processed.
+            if report_timestamp <= last_verified {
```

Comment:
why would this be the case, ie that we will reload reports that were previously verified ?

Reply:
Because we're storing the timestamp of the packet and not the file. Files should always have a later timestamp than the packets they contain